### ISSUE TITLE
Unintended directive ordering

### ISSUE THREAD

**Issue by dhaavi** [mholt/caddy#2647]
While adding my own directive, I noticed that the plugin `s3browser` may be placed incorrectly. As I understand, this plugin also serves content (cached from the proxy directive), but it is placed _before_ the part with all the authentication plugins. I suggest moving it behind them.

On a general note, maybe plugin directives could be grouped by their function. Right now there is no clear place where which plugin should go. Commented sections for "header mangling", "access control", "content" and so forth would be helpful.

https://github.com/mholt/caddy/blob/d11b648137e6c0e3186f46661d145edd2b21b933/caddyhttp/httpserver/plugin.go#L656

**Comment by mholt** [comment]
@techknowlogick - could you take a look at this and see what you think?

**Comment by techknowlogick** [comment]
To clarify the `s3browser` browser plugin only lists the files in the s3 (or equivalent service) bucket, it's the proxy plugin that does the actual serving of data from s3.

If you are relying on an auth plugin to protect your data in terms of this specific plugin, the proxy plugin requires s3 data to be publicly accessible so someone could fetch your data by going directly to s3. Although you could try to use http basic auth to pass credentials to s3 via the proxy plugin, but I haven't tried that so it is a bit experimental.

In terms of moving the directive down I'm ok with that too. If a PR does do that please ping me so I can test the changes to see if the plugin still functions as expected (going through the code in my plugin I see no problems, but verification before the change is made is nice).

As for overall policy of grouping directives by type in the file I can only say that it sounds like a good idea, but as I'm unfamiliar with codebase of most of the plugins I can't say for certain as some may require being in the set order.

Thanks for requesting comment on this :)

**Comment by dhaavi** [comment]
Thanks for clarifying @techknowlogick.

Yes, the proxy plugin fetches the data, but the s3browser plugin may serve it from cache, as explained in your README: `Note: For performance reasons, the file listing is fetched once every 5minutes to reduce load on S3 (or S3 equivalent).`

I would therefore suggest to move it after the authentication plugins.

If desired, I am happy to take a shot at it and create a PR for this.
I'd also try group directives together and comment on these blocks without changing too much.

**Comment by techknowlogick** [comment]
I want to be very clear, s3browser does **not** serve the data in any way, nor does it cache it, it is the proxy plugin that handles the serving of files. The note about caching, is about caching the **list** of files in s3, not the files themselves.

I would like to repeat that moving this plugin below any auth plugin may only give the appearance of security, and could give a false sense to those trying to protect their files.

**Comment by dhaavi** [comment]
I see, thank you for elaborating on this.

On a side note: Generally, I don't think that moving the plugin somewhere else will change the appearance at all, because the average caddy user has no idea about this list. I myself did not know about it for a long time, I actually thought the plugins were loaded in the order I put the in the Caddyfile. And I only discovered the list when I got really confused about some of caddy's behavior and started digging.

If it does not change anything really, then we can leave it as it is, I was just curious because it felt out of place, as the plugin serves content (even if not by itself).
